### ISSUE TITLE
Window-less desktop has wrong focus upon an application's quit

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#375]
I didn't investigate but I wouldn't be surprised if this is the same cause as #371.

Repro steps are easy:
- drag&drop qed.app and some text file onto empty desktop
- drag&drop the text file on qed.app
- close qed.app
- notice that it wasn't the desktop which got topped: it's the XaAES client

As soon as at least one desktop window is open, focus is set correctly.

It had also an interesting evolution:

- 1.16.1 (0.992 from Oct 10 2004): reproducible
- 1.16.3 (0.998 from May 6 2007): **not** reproducible
- 1.17.0 (0.999 from Jan 1 2011): **not** reproducible
- 1.18.0 (1.5.5 from 19.03.2013): reproducible
- latest snapshot: reproducible

So Ozk found and fixed the bug but Helmut reintroduced it again.

**Comment by mikrosk** [comment]
If another application is open, e.g. hypview, it's even funnier: hypview gets topped but at the same time so is the desktop menu.

**Comment by th-otto** [comment]
Imho the behaviour with hypview is correct. Its the only application that has a window open, so it is topped. But it doesn't have a menubar when run as ACC, so the desktop's menubar is displayed.

The initial bug may have to do to do with the setting of `clwtna`

**Comment by mikrosk** [comment]
> Its the only application that has a window open, so it is topped.

While that is true, its initial state was _not_ topped. So from the user perspective, it should return to this state upon qed.app's exit.

> But it doesn't have a menubar when run as ACC, so the desktop's menubar is displayed.

I ran it as .APP (but it's true it doesn't have its own menu)

> The initial bug may have to do with the setting of `clwtna`

All XaAES builds used the same configuration file.

**Comment by th-otto** [comment]
> All XaAES builds used the same configuration file.

Yes, but IIRC that `clwtna` option was introduced by Helmut, that might explain why behaviour is now different.

**Comment by mikrosk** [comment]
> IIRC that clwtna option was introduced by Helmut

No, it was introduced in 100b0977. Similar setting, `next_active`, was also introduced by Ozk.

**Comment by MedourMehdi** [comment]
Hi guys,

Sorry but I can't reproduce

- I dragged and drop qed.app + readme.txt on the desktop
- Then I dragged and dropped readme.txt on qed.app
- I closed qed.app

Everything is working correctly, also with hypview opened first

Does this bug is fixed?

**Comment by mikrosk** [comment]
When you say "correctly", that means what? That a window-less Teradesk was the next active application? I.e. that you see it's menu bar? If so, was it with default settings, no tweaks to xaaes.cnf or Teradesk.inf?

**Comment by MedourMehdi** [comment]
Yes the menu is the previous one - I have tweak the widget and background.

<img width="900" alt="Image" src="https://github.com/user-attachments/assets/ce9533e4-65ea-4234-a93d-859753b525bc" />

**Comment by mikrosk** [comment]
Start with the original scenario - no windows / applications at all.

**Comment by jflemaire** [comment]
Hey! Someone is using GEMClip! Or at least reading the doc ;-) This might motivate me to update it one of these days.

**Comment by MedourMehdi** [comment]
Thanks - I can now reproduce.

I've a problem building the latest repo - xaaes.km give me a black screen - I'll try to figure out why.

**Comment by MedourMehdi** [comment]
> Hey! Someone is using GEMClip! Or at least reading the doc ;-) This might motivate me to update it one of these days.

I personally find it (and gemdict too) to be the most beautiful and best coded application

**Comment by MedourMehdi** [comment]
I have 

`exception 10 for AESSYS pc:10A2D78 addr:0`

Could it be related also to the gcc version?

I'll try to use an older version of freemint to find out if yes

**Comment by mikrosk** [comment]
Currently, xaaes is compiled with gcc 7.5 which obviously works but getting a null pointer with later gcc surely isn't expected...

**Comment by mikrosk** [comment]
Oh wait, you are referring to older xaaes not working too. That sounds like a setup issue, I updated my setup a week ago or so, tested on Falcon, aranym and Hatari.

**Comment by MedourMehdi** [comment]
Yes I just type make and replaced xaaes.km from latest snapshot build with the new built one from compile040 dir...

I'll try with gcc 7.5 when I'll found time to setup such environnement.

**Comment by mikrosk** [comment]
@th-otto do you use your toolchain when playing with freemint or just stick to 7.5?

**Comment by th-otto** [comment]
Depends. Normally, i try to use gcc 7 to be consistent with the snapshots. Sometimes i also use gcc 4. But i think i never used newer gcc, except to check for new warnings. Might as well be that there are some pitfalls (remember the implementation of calloc in mintlib, which caused newer compilers to emit a recursive call to calloc).

**Comment by mikrosk** [comment]
Hmm, so maybe @MedourMehdi did find something fishy with the new compiler, then. I wonder whether there isn't at least a new warning issued or something.

**Comment by th-otto** [comment]
Just tried, and used both a freemint kernel & xaees.km compiled by gcc 14. No problems so far,  especially not a black screen.

BTW, the mint kernel is about 14k larger, and xaaes about 19k. So unless there is a noticable performance boost, i would not recommend to use that version.

**Comment by MedourMehdi** [comment]
Thanks for testing @th-otto -
My gcc version is m68k-atari-mint-gcc (MiNT 20200501) 9.3.1 20200501

I'll test other version to figure what is happening.

I hadn't any warning or error during the build process.

**Comment by xdelatour** [comment]
It seems that only `next_active` is involved when closing an application. And indeed, if there are no open windows, AES becomes active. Perhaps both should be done? Look for the application that has windows, and switch to the previous client if no windows are found (and `next_active` will become useless).

https://github.com/freemint/freemint/blob/08f5b942bfd6e2f7a2fc5deddb34283c9177ce91/xaaes/src.km/xa_appl.c#L714-L732

**Comment by mikrosk** [comment]
> And I'm curious what happens if I set it to client: it is now, in both cases (clean desktop and non-topped hypview), Teradesk and its menu shown?

Answer to this is: yes, it is.
